internal/controller/stages/control_flow_stages.go

```diff
@@ -0,0 +1,603 @@
+package stages
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/controller"
+	rollouts "github.com/akuity/kargo/internal/controller/rollouts/api/v1alpha1"
+	"github.com/akuity/kargo/internal/indexer"
+	"github.com/akuity/kargo/internal/kargo"
+	"github.com/akuity/kargo/internal/kubeclient"
+	libEvent "github.com/akuity/kargo/internal/kubernetes/event"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+type ControlFlowStageReconciler struct {
+	cfg           ReconcilerConfig
+	client        client.Client
+	eventRecorder record.EventRecorder
+}
+
+// NewControlFlowStageReconciler returns a new control flow Stage reconciler.
+// After creating the reconciler, call SetupWithManager to register it with a
+// controller manager.
+func NewControlFlowStageReconciler(
+	cfg ReconcilerConfig,
+) *ControlFlowStageReconciler {
+	return &ControlFlowStageReconciler{
+		cfg: cfg,
+	}
+}
+
+// SetupWithManager sets up the control flow Stage reconciler with the given
+// controller manager. It registers the reconciler with the manager and sets up
+// watches on the required objects.
+func (r *ControlFlowStageReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
```

Comment:
Even if I like to be idiomatic and puristic at times, this is how controller-runtime projects do it — and would fit more into what we have set out for ourselves with #1479.

https://xkcd.com/927/

Reply:
Far from being a hill I would die on, I think the advantage of the other approach was that a caller in another package could obtain a fully-initialized reconciler after a single function call. With this approach, it requires two, which creates the possibility of obtaining a _partially-initialized_ reconciler.

As I said, it's not a hill I would die on. Merely explaining the rationale behind the prior approach.